devito/function.py

```diff
@@ -259,43 +291,151 @@ def _indices(cls, **kwargs):
     def shape(self):
-        return self.shape_data
+        """
+        Shape of the domain associated with this :class:`Function`.
+        The domain constitutes the area of the data written to in a
+        stencil update and excludes the read-only stencil boundary.
+        """
+        return self.shape_domain
+
+    @property
+    def shape_domain(self):
+        """
+        Shape of the domain associated with this :class:`Function`.
+        The domain constitutes the area of the data written to in a
+        stencil update and excludes the read-only stencil boundary.
+
+        .. note::
+
+            Alias to ``self.shape``.
+        """
+        return tuple(i - j for i, j in zip(self._grid_shape_domain, self.staggered))
+
+    @property
+    def shape_with_halo(self):
+        """
+        Shape of the domain plus the read-only stencil boundary associated
+        with this :class:`Function`.
+        """
+        return tuple(j + i + k for i, (j, k) in zip(self.shape_domain, self._halo))
+
+    @property
+    def shape_allocated(self):
+        """
+        Shape of the allocated data associated with this :class:`Function`.
+        It includes the domain and halo regions, as well as any additional
+        padding outside of the halo.
+        """
+        return tuple(j + i + k for i, (j, k) in zip(self.shape_with_halo, self._padding))
 
     @property
     def space_dimensions(self):
         """Tuple of index dimensions that define physical space."""
         return tuple(d for d in self.indices if d.is_Space)
 
-    def _allocate_memory(self):
-        """Allocate memory in terms of numpy ndarrays."""
-        debug("Allocating memory for %s (%s)" % (self.name, str(self.shape)))
-        self._data_object = Data(self.shape, self.indices, self.space_order, self.dtype)
-        if self._first_touch:
-            first_touch(self)
-        else:
-            self.data.fill(0)
-
     @property
     def data(self):
-        """The value of the data object, as a :class:`numpy.ndarray` storing
-        elements in the classical row-major storage layout."""
-        if self._data_object is None:
-            self._allocate_memory()
-        return self._data_object
+        """
+        The domain data values, as a :class:`numpy.ndarray`.
 
-    def initialize(self):
-        """Apply the data initilisation function, if it is not None."""
-        if self.initializer is not None:
-            self.initializer(self.data)
+        Elements are stored in row-major format.
+        """
+        return self.data_domain
+
+    @property
+    @_allocate_memory
+    def data_domain(self):
+        """
+        The domain data values, as a :class:`numpy.ndarray`.
+
+        Elements are stored in row-major format.
+
+        .. note::
+
+            Alias to ``self.data``.
+        """
+        # TODO: for the domain-allocation switch, this needs to be turned
+        # into a view of the domain region
+        return self._data
```

Comment:
This seems incongruent; it looks here like the switch has already happened, although `self.data_allocated` points to the same object. In contrast, the `self.shape_domain` and `self.shape_allocated` do not return the same values, so one of the two cannot fit the returned `._data` object (missing test?).

Should we maybe error out on all "domain" properties (shape/data) and default to "allocated" throughout until we switch completely?

Reply:
oh yes, I only "anticipated" the implementation of ``shape_allocated`` etc assuming nobody will use them until the switch. I can definitely make everything more congruent, I think

Reply:
so here's my final approach: ATM, "everything is domain". Trying to access halo/padding properties will raise a ``NotImplementedError``. The infrastructure for the switch is there though (suitably commented out and with useful docstrings)

Reply:
Great, thanks. :+1: